keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,368 @@
+in default case, data network connectivity is offered via normal 1 Gbps NICs.<br/>
+**Motivation:** If some workloads demand higher network bandwidth, it should be possible to run these workloads on selected nodes.<br/>
+**Can this be solved without resource classes:** Taints and tolerations can help in steering pods but the problem in that there is no way today to have access control over use of tolerations and therefore if multiple users are there, it is not possible to have control on allowed tolerations.<br/>
+**How Resource classes can solve this:** I can define a ResourceClass for the high-performance NIC with minimum bandwidth requirements, and makes sure only users with proper quota can use such resources.
+
+- I want to be able to utilize different 'types' of a HW resource while not losing workload portability when moving from one cluster to another. There can be Nvidia GPUs on one cluster and AMD GPUs on another cluster. This is example of different ‘types’ of a HW resource(GPU). I want to offer GPUs to be consumed under a same portable name, as long as their capabilities are almost same. If pods are consuming these GPUs with a generic resource class name, workload can be migrated from one cluster to another transparently.<br/>
+**Motivation:**  less downtime, optimal use of resources<br/>
+**How Resource classes can solve this:** Explained above<br/>
+**Can this be solved without resource classes:** No
+
+### As a developer:
+- I want the ability to be able to request devices which have specific capabilities. Eg: GPUs that are Volta or newer.<br/>
+  **Motivation:** I want minimum guaranteed compute performance<br/>
+  **Can this be solved without resource classes:**<br/>
+  - Yes, using node labels and NodeLabelSelectors.
+    Problem: Same problem of lack of access control on using labelselectors at user level as with the use of tolerations.
+  - OR, Instead of using resource class, provide flexibility to query resource properties directly in pod container resource requests.
+    Problem: In a large cluster, computing operators like “greater than”, “less than” at pod creation can be a very slow operation and is not scalable.
+
+  **How Resource classes can solve this:**
+The Kubernetes scheduler is the central place to map container resource requests expressed through ResourceClass names to the underlying qualified physical resources, which automatically supports metadata aware resource scheduling.
+
+- As a data scientist, I want my workloads to use advanced compute resources available in the running clusters without understanding the underlying hardware configuration details. I want the same workload to run on either on-prem Kubernetes clusters or on cloud, without changing its pod spec. When a new hardware driver comes out, I hope all the required resource configurations are handled properly by my cluster operators and things will just continue to work for any of my existing workloads.<br/>
```

Comment:
This use case seems too vague. TBH, the workloads that consume additional HW are specialized enough that it requires developer maintenance and cluster admins may not be able to homogenize different environments.

Reply:
I tend to disagree. I think a big value of Kubernetes is to allow separation of concern that application developers can focus on their own software with underlying infrastructure taken cared by cluster admins.

Reply:
@vishh quoting Henry from Ebay in the support of workload portability: 
>HI Vish, Bobby, it is not exactly this requirement. Currently we ask developers to submit resource specifications for GPU using name of the cards to our data center :

>"accelerator": {
"type": "gpu",
"quantity": "1",
"labels": {
"product": "nvidia",
"family": "tesla",
"model": "m40"
}
}
>
>But when we go to other cloud such as Google or AWS they may not have the same cards.

>So I was wondering if we could offer resource such as CUDA cores and memory as resource specifications rather actual name and type of the cards.
>
>However, different cards such as AMD vs NVIDIA was not the goal because we know program code against NVIDIA cards will not work well if run with AMD cards.

Reply:
I'll wait for Henry to respond to that thread. I think you all should solicit feedback from real users (I'm thinking ML WG, SIG BIG Data, etc.) to figure out if this is really feasible. No user that I have spoken to is ready to consume this level of sophistication today.

Reply:
I think we are proposing an infrastructure building block here whose focusing users are infrastructure admins and developers who want to make their systems easier to use by hiding the underlying hardware details from end users.

Reply:
> As a data scientist, I want my workloads to use advanced compute resources available in the running clusters without understanding the underlying hardware configuration details.

Yes and no. But mostly no. As mentioned by @vishh this is too vague and a lot of things that this statements covers is outside of K8s' scope.
- It might make sense when your users only request one GPU (but in that case does ECC on/off is a HW config?)
- When a user requests more than one GPU users should at the very least be able to specify if GPUs are linked through NVLINK


> I want the same workload to run on either on-prem Kubernetes clusters or on cloud, without changing its pod spec. 

What's blocking this today or what might be blocking this in the future?

> When a new hardware driver comes out, I hope all the required resource configurations are handled properly by my cluster operators and things will just continue to work for any of my existing workloads.

How is this in the scope of Resource Classes